gpui: Implement support for wlr layer shell

happenslol opened the pull request:
This reintroduces `layer_shell` support after #32651 was reverted. On top of that, it allows setting options for the created surface, restricts the enum variant to the `wayland` feature, and adds an example that renders a clock widget using the protocol.

I've renamed the `WindowKind` variant to `LayerShell` from `Overlay`, since the protocol can also be used to render wallpapers and such, which doesn't really fit with the word.

Things I'm still unsure of:
- We need to get the layer options types to the user somehow, but nothing from the `platform::linux` crate was exported, I'm assuming intentionally. I've kept the types inside the module (instead of doing `pub use layer_shell::*` to not pollute the global namespace with generic words like `Anchor` or `Layer` Let me know if you want to do this differently.
- I've added the options to the `WindowKind` variant. That's the only clean way I see to supply them when the window is created. This makes the kind no longer implement `Copy`.
- The options don't have setter methods yet and can only be defined on window creation. We'd have to make fallible functions for setting them, which only work if the underlying surface is a `layer_shell` surface. That feels un-rust-y.

CC @zeroeightysix  
Thanks to @wuliuqii, whose layer-shell implementation I've also looked at while putting this together.

Release Notes:

- Add support for the `layer_shell` protocol on wayland

happenslol committed:
gpui: Implement support for wlr layer shell

Co-authored-by: Ridan Vandenbergh <ridanvandenbergh@gmail.com>

Guibi1 commented:
This would be amazing to see merged. +1

mikayla-maki reviewed (dismissed):
Hello! Thanks for making this PR.

Can you make a quick GPUI example demonstrating the behavior for this feature? As well as fix the merge conflicts?

With those two done, we can test this and get it merged :)

cole-miller reviewed (changes requested):
Thanks, a couple of feedback items but generally this looks great! Also was able to run the example smoothly, thanks for adding that.

happenslol commented on crates/gpui/src/platform/linux/wayland/window.rs:
```diff
@@ -282,31 +348,91 @@ impl WaylandWindow {
-        }
+        let surface_state = match (&params.kind, globals.layer_shell.as_ref()) {
+            // By matching on layer_shell here, we can fall back to creating a normal xdg_surface
+            // if the user requested a LayerShell but the compositor doesn't support the protocol.
```
Yeah, that makes sense to me. What kind of error should be returned here? Could you point me to an example in the codebase where a match-able error is returned so I can use that as a template?

happenslol commented:
Thanks for the review! I think I'll get around to this on the weekend :slightly_smiling_face:

happenslol committed:
Return error when layer_shell protocol is not supported

Instead of falling back to a normal xdg_surface, this will abort window
creation if layer_shell is not supported. This also cleans up surface
state creation a bit by pulling it out into a separate function.

happenslol committed:
Document order of destroy events when dropping wayland windows

Seems like the current order would have been fine, but I've added
comments and links to provide more clarity in case something changes in
the future here.

happenslol committed:
Update max layer_shell protocol version

The protocol version was falsely pinned to 1, which caused some features
such as layer selection and keyboard interactivity not to work correctly
on all compositors.

cole-miller reviewed (changes requested):
Thanks for addressing feedback, the code itself looks good, but could you undo the reordering of dependencies in gpui/Cargo.toml? (They should be sorted alphabetically.) That should also help resolve the merge conflict in that file. Other than that I think this is ready to merge!

happenslol committed:
Fix Cargo.toml order and make rebased code compatible

happenslol committed:
Return an error struct for missing layer_shell support

cole-miller reviewed (approved):
Thanks! We use squash and merge so no need to squash manually.
